service/role.go

```diff
@@ -145,109 +215,174 @@ func (r *roleService) getRoleToken(ctx context.Context, domain, role, proxyForPr
 	return tok, nil
 }
 
-// handleExpiredHook is a handler function for gache expired hook
-func (r *roleService) handleExpiredHook(fctx context.Context, key string) {
-	domain, role := decode(key)
-	r.updateRoleToken(fctx, domain, role, "", r.expiry, r.expiry)
-}
+// refreshRoleTokenCache returns the error channel when it is updated.
+func (r *roleService) RefreshRoleTokenCache(ctx context.Context) <-chan error {
+	glg.Info("refreshRoleTokenCache started")
 
-// updateRoleToken returns RoleToken struct or error.
-// This function ask athenz to generate role token and return, or return any error when generating the role token.
-func (r *roleService) updateRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	echan := make(chan error)
+	go func() {
+		defer close(echan)
 
-	tok, err, _ := r.group.Do(encode(domain, role), func() (interface{}, error) {
-		// get the role token
-		tok, err := r.token()
-		if err != nil {
-			return nil, err
-		}
+		r.domainRoleCache.Foreach(ctx, func(key string, val interface{}, exp int64) bool {
+			domain, role, principal := decode(key)
+			for err := range r.updateRoleTokenWithRetry(ctx, domain, role, principal, r.expiry, r.expiry) {
+				echan <- err
+			}
+			return true
+		})
+	}()
 
-		// concat URL string and url parameters
-		u := getRoleTokenAthenzURL(r.athenzURL, domain, role, minExpiry, maxExpiry, proxyForPrincipal)
+	return echan
+}
 
-		req, err := http.NewRequest(http.MethodGet, u, nil)
-		if err != nil {
-			return nil, err
-		}
+// handleExpiredHook is a handler function for gache expired hook
+func (r *roleService) handleExpiredHook(ctx context.Context, key string) {
+	glg.Debugf("handleExpiredHook(), key: %s", key)
```

Comment:
should this be `Warn` level?
There is a time gap between delete and fetch, and we do not expect this to run.

Although sidecar will ask ZTS directly on cache miss, it is still better to warn in the log.

Reply:
@kevindiu ping